### ISSUE TITLE
Prototype performance portability

### ISSUE THREAD

**Issue by sethrj** [celeritas-project/celeritas#283]
Do an initial port of enough core Celeritas components to at least run some demo applications on non-CUDA hardware.

**Comment by drbenmorgan** [comment]
@sethrj, let me know if there's anything I can do here - I can see an overlap with #286 on the AdePT side.

**Comment by sethrj** [comment]
Thanks @drbenmorgan ! One of our first burning questions is "what RNGs work portably with the new thread parallelism?"

Tom and my plan is:
- Install NVHPC on our test system
- Try a dead-simple parallel-for-offload code with the C++ pseudorandom number generators (MT?) to see if it compiles and produces the correct results (end-of-november timeframe)
- Add a CMake configure switch for targeting NVC++ offloading with UVM
- Port the KN demo kernel to use parallel algorithms (end-of-december timeframe)

If you're up for playing around with the NVHPC features we'd be glad to have another hand! We can coordinate here, or over email, or at our next collaborative meeting.

**Comment by drbenmorgan** [comment]
Very happy to lend a hand here! Fine to coordinate however suits you and Tom best - just to get started, which version of NVHPC/CUDA and underlying GCC are you going to target? I can then get that set up on my local spack instance so we're using the same base.

**Comment by sethrj** [comment]
That would be great! Since we're acknowledging that this is a "bleeding edge" capability I think @tmdelellis and I are OK with trying out the latest version of the compilers, under the assumption that by the time we actually get a working code the production systems will have that version installed. So I guess CUDA 11.5 and GCC 11?

**Comment by drbenmorgan** [comment]
O.k., installing `spack install nvhpc` on a Linux CentOS7 system with spack GCC 11.2.0 seems to be fine on this side:

```
$ spack load gcc@11.2.0 nvhpc
$ nvc++ --version

nvc++ 21.9-0 64-bit target on x86-64 Linux -tp haswell 
NVIDIA Compilers and Tools
Copyright (c) 2021, NVIDIA CORPORATION & AFFILIATES.  All rights reserved.

$ nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2021 NVIDIA Corporation
Built on Wed_Jul_14_19:41:19_PDT_2021
Cuda compilation tools, release 11.4, V11.4.100
Build cuda_11.4.r11.4/compiler.30188945_0
$
```

So CUDA 11.4 as supplied in the latest version from spack, but likely close enough (not sure why it'. A couple of trivial codes also compile and run.

**Comment by brycelelbach** [comment]
Can someone point me at the source code for the Klein-Nishina photon transport mini app?

**Comment by sethrj** [comment]
@brycelelbach (cross-posted from email):

Certainly. The main outer loop that I walked through today is https://github.com/celeritas-project/celeritas/blob/master/app/demo-interactor/KNDemoRunner.cc . There are a few abstractions in place to make it easier to allow our code to run as a manual-memory-management CUDA code while still supporting running on the CPU with standard C++, so you might find the code easier to tinker with if you revert to the early-February version of the code (22b99bdded).

Our PRNG requirement at present is to have one independent stream per thread. The only "communication" is when we calculate per-thread seeds on the CPU at the beginning of the run. (Using random seeds is of course only an *approximation* to independent streams, but we've found it works well in our other nuclear application. Guaranteeing fully independent streams is difficult if you don't know the RNG usage per thread a priori.) We retain a state for each thread in global memory. (In one iteration of our code, NVCC's optimizer was smart enough to actually pull the state into local memory during the kernel in which it was used and write it out at the end.)
